### DISCUSSION TITLE
Extension clean-up after experiment

### DISCUSSION THREAD

**Discussion by mgrewe** [jspsych/jsPsych#2809] [Q&A]
Hi all,

I'm currently busy with implementing a unity extension that allows inclusion of the WebGL player in jsPsych. While it can already load and display the content, I get an error on finish of the experiment. The reason is, that the canvas is destroyed, that Unity is attached to. Unfortunately, I did not found a way to attach an "on_finish" (experiment, not trial) callback from within the extension.

Is there any event that I can listen to? Would it be otherwise possible to allow adding callback functions from within the extension for any necessary clean-up work?

Thanks a lot,
Martin

**Comment by jodeleeuw** [comment]
Hi @mgrewe,

Could you say a bit more about why this is an extension and not a plugin? The idea with extensions was that they can add functionality to any existing plugin. Adding a canvas to the screen would probably not be compatible with existing plugins, but maybe I'm misunderstanding the idea here.

Thanks!

**Reply by mgrewe** [reply]
Can you shortly elaborate on the "destruction" problem? Is there a way to gently close unity after the experiment had been finished, either with an extension or preload plugin?

Thanks

**Comment by mgrewe** [comment]
Hi @jodeleeuw,
thanks for your answer. I absolutely agree that plugins would be the usual way. However, AFAIK, the plugin mechanism does not allow providing parameters to the constructor (e.g., the URL of the loader script, data, WASM etc.) , which would be important to do the heavy preloading of the WebGL player (usually >> 5MB) already during loading of the experiment, and not directly before the actual trial. I currently planned to also provide a unity plugin that is then responsible for insertion of the canvas element into the DOM tree. Hope this makes sense.
Cheers

**Reply by jodeleeuw** [reply]
Got it. Would it make sense to create a plugin that preloads the WebGL player for use by a different plugin, in the way that the existing preload plugin handles preloading of media assets that are then used by other plugins later? This might give developers more options, e.g., you could orient users to the experiment before doing a long content load.

**Reply by mgrewe** [reply]
Thanks for the pointer. Will have a look at it tomorrow.

**Reply by mgrewe** [reply]
BTW, I just recognized that an extension would make sense because it can be used by several plugins which do a lightweight configuration of the different conditions of the paradigm. E.g., we have sync/async conditions in a virtual mirror experiment. By providing the condition as a param to to the plugin, it can forward them to the unity app, which implements the basic paradigm and just adjusts the synchrony of the movements.

**Reply by mgrewe** [reply]
Hey @jodeleeuw 

I had a look at the Preload plugin. Am I correct that it only handles preloading of Image/Video/Audio? Is there any chance to also allow preloading of arbitrary binary blobs (e.g., OBJECTS)?

Thanks
Martin
